File: backend/mscliente/src/main/java/mscliente/mscliente/listeners/ClienteListener.java
package mscliente.mscliente.listeners;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import mscliente.mscliente.DTO.ClienteDTO;
import mscliente.mscliente.DTO.ReservaCreationResponseDTO;
import mscliente.mscliente.DTO.ReservaInputDTO;
import mscliente.mscliente.DTO.ReservaOutputDTO;
import mscliente.mscliente.services.ClienteService;
import mscliente.mscliente.services.MilhasService;
import org.springframework.amqp.AmqpRejectAndDontRequeueException;
import org.springframework.amqp.rabbit.annotation.RabbitListener;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class ClienteListener {

    @Autowired
    private ClienteService clienteService;

    @Autowired
    private MilhasService milhasService;

    @Autowired
    private ObjectMapper objectMapper;

    @RabbitListener(queues = "autocadastro.cliente")
    public String receberAutocadastro(String clienteJson) {
        try {
            ClienteDTO clienteDTO = objectMapper.readValue(clienteJson, ClienteDTO.class);
            ClienteDTO createdClient = clienteService.saveCliente(clienteDTO);
            return objectMapper.writeValueAsString(createdClient);
        } catch (IllegalArgumentException e) {
            return "{\"erro\":\"CPF ou e-mail já cadastrados.\"}";
        } catch (Exception e) {
            System.err.println("Erro ao salvar cliente: " + e.getMessage());
            throw new AmqpRejectAndDontRequeueException("Erro no autocadastro", e);
        }
    }

    @RabbitListener(queues = "login.cliente")
    public String dadosLoginCliente(String email) {
        try {
            Object dadosCliente = clienteService.findByEmail(email);
            return objectMapper.writeValueAsString(dadosCliente);
        } catch (Exception e) {
            System.err.println("Erro ao buscar: " + e.getMessage());
            throw new AmqpRejectAndDontRequeueException("Erro no login", e);
        }
    }

    @RabbitListener(queues = "criareserva.saldo")
    public String consultaSaldo(String payload) {
        try {
            ReservaInputDTO dadosReserva = objectMapper.readValue(payload, ReservaInputDTO.class);
            ClienteDTO dadosCliente = clienteService.findById(dadosReserva.getCodigo_cliente());

            if (dadosReserva.getMilhas_utilizadas() > dadosCliente.getSaldo_milhas()) {
                throw new IllegalArgumentException("Saldo de milhas insuficiente");
            }

            return objectMapper.writeValueAsString(dadosCliente);
        } catch (Exception e) {
            throw new AmqpRejectAndDontRequeueException("Erro na verificação de saldo", e);
        }
    }

    @RabbitListener(queues = "criareserva.cliente")
    public String efetuarReserva(String payload) {
        try {
            ReservaCreationResponseDTO transaction = objectMapper.readValue(payload, ReservaCreationResponseDTO.class);
            Object result = milhasService.registrarReserva(transaction);
            return objectMapper.writeValueAsString(result);
        } catch (Exception e) {
            throw new AmqpRejectAndDontRequeueException("Erro ao efetuar reserva", e);
        }
    }

    @RabbitListener(queues = "cancelareserva.cliente")
    public String cancelarReserva(String payload) {
        try {
            ReservaOutputDTO reserva = objectMapper.readValue(payload, ReservaOutputDTO.class);
            Object result = milhasService.reembolsarReserva(reserva);
            return objectMapper.writeValueAsString(result);
        } catch (Exception e) {
            throw new AmqpRejectAndDontRequeueException("Erro ao cancelar reserva", e);
        }
    }

    // @RabbitListener(queues = "cancelavoo.cliente")
    // public String canceladoVoo(String payload) {
    //     try {
    //         List<ReservaOutputDTO> reservas = objectMapper.readValue(
    //                 payload,
    //                 new TypeReference<List<ReservaOutputDTO>>() {
    //                 }
    //         );
    //         Object result = milhasService.reembolsarVoo(reservas);
    //         return objectMapper.writeValueAsString(new RabbitMessageDTO(true, result));
    //     } catch (Exception e) {
    //         throw new AmqpRejectAndDontRequeueException("Erro ao cancelar voo", e);
    //     }
    // }
}

File: backend/cliente-service/src/main/kotlin/br/ufpr/dac/cliente_service/resource/ClienteListener.kt
package br.ufpr.dac.cliente_service.resource

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import jakarta.validation.ConstraintViolation
import jakarta.validation.Validation
import jakarta.validation.Validator
import jakarta.validation.ValidatorFactory
import org.springframework.amqp.rabbit.annotation.RabbitListener
import org.springframework.stereotype.Service
import utils.dto.*
import utils.gson.ZonedDateTimeAdapter
import java.lang.IllegalArgumentException
import java.time.ZonedDateTime

@Service
class ClienteListener(private val service: ClienteService, private val milhasService: MilhasService) {
    private final val gson: Gson = GsonBuilder()
        .registerTypeAdapter(ZonedDateTime::class.java, ZonedDateTimeAdapter())
        .create()
    private final val factory: ValidatorFactory = Validation.buildDefaultValidatorFactory()
    private final val validator: Validator = factory.validator

    @RabbitListener(queues = ["emiratads.autocadastro.cliente"], errorHandler = "customErrorHandler")
    fun autocadastroSaga(obj: String): String {
        val response: RabbitMessageDTO<ClienteOutputDTO>
        val clienteInput = gson.fromJson(obj, ClienteInputDTO::class.java)
        val violations: Set<ConstraintViolation<ClienteInputDTO>> = validator.validate(clienteInput)

        if (violations.isEmpty()) {
            val cliente = service.saveCliente(clienteInput)
            response = RabbitMessageDTO(true, cliente)
        } else {
            response = RabbitMessageDTO(
                false,
                "Cliente enviado com formato inválido: $violations",
                "ConstraintViolationException"
            )
        }

        return gson.toJson(response)
    }

    @RabbitListener(queues = ["emiratads.login.cliente"], errorHandler = "customErrorHandler")
    fun dadosLoginCliente(code: String): String {
        val idCliente = code.toLong()
        val dadosCliente = service.getClienteByID(idCliente)

        return gson.toJson(dadosCliente)
    }

    @RabbitListener(queues = ["emiratads.criareserva.saldo"], errorHandler = "customErrorHandler")
    fun consultaSaldo(payload: String): String {
        val dadosReserva = gson.fromJson(payload, ReservaInputDTO::class.java)
        val dadosCliente = service.getClienteByID(dadosReserva.codigo_cliente)

        if (dadosReserva.milhas_utilizadas > dadosCliente.saldo_milhas){
            throw IllegalArgumentException("Saldo de milhas insuficiente")
        }

        return gson.toJson(RabbitMessageDTO(true, dadosCliente))
    }

    @RabbitListener(queues = ["emiratads.criareserva.cliente"], errorHandler = "customErrorHandler")
    fun efetuarReserva(payload: String): String {
        val transaction = gson.fromJson(payload, ReservaCreationResponseDTO::class.java)
        val result = milhasService.registrarReserva(transaction)

        return gson.toJson(RabbitMessageDTO(true, result))
    }

    @RabbitListener(queues = ["emiratads.cancelareserva.cliente"], errorHandler = "customErrorHandler")
    fun cancelarReserva(payload: String): String {
        val reserva = gson.fromJson(payload, ReservaOutputDTO::class.java)
        val result = milhasService.reembolsarReserva(reserva)

        return gson.toJson(RabbitMessageDTO(true, result))
    }

    @RabbitListener(queues = ["emiratads.cancelavoo.cliente"], errorHandler = "customErrorHandler")
    fun canceladoVoo(payload: String): String {
        val reservas: List<ReservaOutputDTO> = gson.fromJson(
            payload,
            object : TypeToken<List<ReservaOutputDTO>>() {}.type
        )
        val result = milhasService.reembolsarVoo(reservas)
        return gson.toJson(RabbitMessageDTO(true, result))
    }
}